### PR TITLE
Enable ``3.11`` in the CI

### PR DESCRIPTION
**Pull request by DanielNoord**
]
## Steps

- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Related Issue

Ref. #1516

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -81,9 +81,8 @@ jobs:
     runs-on: ubuntu-latest
     timeout-minutes: 20
     strategy:
-      fail-fast: true
       matrix:
-        python-version: [3.6, 3.7, 3.8, 3.9, "3.10"]
+        python-version: [3.6, 3.7, 3.8, 3.9, "3.10", "3.11.0-alpha.7 - 3.11.0"]
```
Comment (cdce8p) [inline .github/workflows/ci.yaml:85]: Why use `"3.11.0-alpha.7 - 3.11.0"`? Maybe `3.11-dev` would work?

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -179,11 +179,12 @@ jobs:
     name: tests / run / ${{ matrix.python-version }} / Windows
     runs-on: windows-latest
     timeout-minutes: 20
+    if: ${{ always() }}
```
Comment (cdce8p) [inline .github/workflows/ci.yaml:182]: That doesn't look great.

**Review by cdce8p — COMMENTED** [review]
Python 3.11 is still in alpha at this time. Maybe it's better to add it as a separate job for now, instead of using `always()`. We can merge that back once the beta has gotten stable enough.

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (DanielNoord) [reply .github/workflows/ci.yaml:85]: This was just to make sure we would get the latest version (which was necessary for `tomllib` for the `pylint` counter-PR). I'll try with `3.11-dev`.

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (DanielNoord) [reply .github/workflows/ci.yaml:182]: I'll create a new job.

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -130,11 +130,65 @@ jobs:
...
+          name: coverage-${{ matrix.python-version }}
+          path: .coverage
+
   coverage:
     name: tests / process / coverage
     runs-on: ubuntu-latest
     timeout-minutes: 5
-    needs: ["tests-linux"]
+    needs: ["tests-linux", "tests-linux-alpha"]
+    if: ${{ always() }}
```
Comment (DanielNoord) [inline .github/workflows/ci.yaml:191]: @cdce8p This remains necessary for the coverage to include `3.11` which would be good for any changes we introduce to fix regression with that version. I have removed it from the windows job though.

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -179,6 +233,7 @@ jobs:
     name: tests / run / ${{ matrix.python-version }} / Windows
     runs-on: windows-latest
     timeout-minutes: 20
+
```
Comment (DanielNoord) [inline .github/workflows/ci.yaml:236]:
```suggestion
```

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -130,11 +130,65 @@ jobs:
...
 
+  tests-linux-alpha:
+    name: tests / run / ${{ matrix.python-version }} / Linux
+    runs-on: ubuntu-latest
+    timeout-minutes: 20
+    strategy:
+      matrix:
+        python-version: ["3.11-dev"]
+    outputs:
+      python-key: ${{ steps.generate-python-key.outputs.key }}
```
Comment (cdce8p) [inline .github/workflows/ci.yaml:141]: The job `output` isn't actually used anywhere. So this is safe to remove.
`${{ steps.python.outputs.python-version }}` does work without it.

File: .github/workflows/ci.yaml [hunk 6]
```diff
@@ -130,11 +130,65 @@ jobs:
...
+    steps:
+      - name: Check out code from GitHub
+        uses: actions/checkout@v3.0.1
+        with:
+          fetch-depth: 0
```
Comment (cdce8p) [inline .github/workflows/ci.yaml:146]: Why set it to `0`? That will fetch the complete history which is unnecessary. A shallow copy (with the default `1`) should be enough. https://github.com/actions/checkout#usage

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (cdce8p) [reply .github/workflows/ci.yaml:191]: Not sure that is all that useful. The coverage is merged anyway, so the only thing we would get would be the lines with `PY311_PLUS`.

File: .github/workflows/ci.yaml [hunk 7]
```diff
@@ -227,6 +281,59 @@ jobs:
           . venv\\Scripts\\activate
           pytest tests/
 
+  tests-windows-alpha:
+    name: tests / run / ${{ matrix.python-version }} / Windows
+    runs-on: windows-latest
```
Comment (cdce8p) [inline .github/workflows/ci.yaml:286]: Do we want / need to test Windows? For almost everything the `Linux` test should be enough. Once merged we'll test Windows anyway.

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (DanielNoord) [reply .github/workflows/ci.yaml:191]: I think that would still be helpful. For example, more "risky" PRs such as https://github.com/PyCQA/astroid/pull/1513 would probably be less risky if we know coverage is okay.

**Review by cdce8p — APPROVED** [review]
Pushed some minor changes. Looks good now!

**Pull request merged**